### PR TITLE
Make vehicle wheels steer smoothly

### PR DESCRIPTION
**Pull request by drummyfish**
Hello, this is my first attempt at PR to this repo. Feel free to throw criticism at me :)

EDIT:

Unlike in vanilla engine, the vehicle wheels instantly turned to their limit angle when the player steered. I corrected this by smoothly changing the angle in each physics tick, independently of framerate. It's not just a change of animation but actual physics behavior that allows for finer steering.

### CODE REVIEW

**Comment by ShFil119** [comment]
I'll give you an advice, which I got. ;)
"Thank you for your contribution, @drummyfish!

I'd like to note that it would be much easier to manage the project and work with PRs if you could provide more definitive commit messages and PR titles.

There's an article that has once helped me understand the importance of being as verbose as possible while working with git, and I'd like to share it with you as I see your contributions being valuable, but the commit messages creating a bottleneck in reviewing and merging them in because it's a big project and it's not easy to maintain.

Here's the link to the article: https://chris.beams.io/posts/git-commit/" ;)

**Comment by drummyfish** [comment]
Thank you! I try to be reasonable with the comments, but I'm still a newbie to contributing, so I will definitely read it. I noticed the code is quite nicely commented compared to some other projects, so I guess you emphasize good documentation, which is definitely good.

**Comment by ShFil119** [comment]
Maintainers are just "picky". So if you want to make your code merged, you should try each available method to satisfy them. 

Good documentation is one best strategy. ;)

File: rwengine/src/objects/VehicleObject.cpp [hunk 1]
```diff
@@ -262,14 +262,23 @@ void VehicleObject::tickPhysics(float dt) {
...
-                float sign = std::signbit(steerAngle) ? -1.f : 1.f;
-                physVehicle->setSteeringValue(
-                    std::min(glm::radians(info->handling.steeringLock),
-                             std::abs(steerAngle)) *
-                        sign,
-                    w);
-                // physVehicle->setSteeringValue(std::min(3.141f/2.f,
-                // std::abs(steerAngle)) * sign, w);
+                float currentVal = physVehicle->getSteeringValue(w);
+                float currentSign = std::signbit(currentVal) ? -1.f : 1.f;
```
Comment (CookiePLMonster) [inline rwengine/src/objects/VehicleObject.cpp]: `std::copysign(1.0f, currentVal)` may be handy here.

**Comment by darkf** [comment]
Anyone tested this yet?

**Comment by darkf** [comment]
I suppose I'll ask again -- @ShFil119, @husho, et al -- anyone up for testing and reviewing this?

**Comment by husho** [comment]
I have tested it on master and it works fine.

File: rwengine/src/objects/VehicleObject.cpp [hunk 2]
```diff
@@ -262,14 +262,23 @@ void VehicleObject::tickPhysics(float dt) {
...
+                float newVal;
+
+                if (std::abs(steerAngle) < 0.001f) {   // no steering?
+                    newVal = std::max(0.0f,std::abs(currentVal) - 4.f * dt) *
+                        currentSign; 
+                } else {
+                    newVal = currentVal + steerAngle * dt * 4.f;
```
Comment (husho) [inline rwengine/src/objects/VehicleObject.cpp]: Not sure if this is the right step value. My guess (from quick glance at decompiled code) it's  ```1 / 0.35 = 2.85714285714 ```

File: rwengine/src/objects/VehicleObject.cpp (hunk 1 above)
Comment (JayFoxRox) [reply rwengine/src/objects/VehicleObject.cpp]: Good suggestion; nit.

**Comment by JayFoxRox** [comment]
I think this is fine as a temporary solution, but it feels slightly hacky / overcomplicated.

LGTM if @drummyfish is not around to address these issues.
In that case, an issue should be created to address some of the review issues (fine-tuning the interpolation, improving code quality)

File: rwengine/src/objects/VehicleObject.cpp [hunk 3]
```diff
@@ -262,14 +262,23 @@ void VehicleObject::tickPhysics(float dt) {
...
+
+                    float limit = glm::radians(info->handling.steeringLock);
+
+                    if (std::abs(newVal) > limit)
+                        newVal = limit * currentSign;
```
Comment (JayFoxRox) [inline rwengine/src/objects/VehicleObject.cpp]: nit: [I like curly braces](https://nakedsecurity.sophos.com/2014/02/24/anatomy-of-a-goto-fail-apples-ssl-bug-explained-plus-an-unofficial-patch/). The google code style doesn't care, so this is a personal opinion.

**Comment by dracc** [comment]
I'd be happy addressing these points and filing a 'fix' PR.

File: rwengine/src/objects/VehicleObject.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dracc) [reply rwengine/src/objects/VehicleObject.cpp]: Would you mind expand on how that would be accomplished? I don't see the case you're thinking about. :(

**Pull request merged**